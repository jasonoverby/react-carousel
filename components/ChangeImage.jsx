import React from 'react';
import PropTypes from 'prop-types';

import { updateImage, getNewImageIndex } from '../lib/update-image';

export const UpdateImage = ({
  incrementOrDecrement,
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => (
  <button
    aria-label={`show ${
      incrementOrDecrement === 'decrement' ? 'previous' : 'next'
    } image`}
    className="changeImage"
    onClick={async () => {
      const updatedImageIndex = getNewImageIndex(
        imageIndex,
        incrementOrDecrement,
      );
      await updateImage({
        setActiveDotIndex,
        setImageIndex,
        setImageLoaded,
        updatedImageIndex,
      });
    }}
    type="button"
  >
    <span className="arrow">
      {incrementOrDecrement === 'decrement' ? '◀' : '▶'}
    </span>
  </button>
);

export const DecrementImage = ({
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) =>
  UpdateImage({
    imageIndex,
    incrementOrDecrement: 'decrement',
    setActiveDotIndex,
    setImageIndex,
    setImageLoaded,
  });

export const IncrementImage = ({
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) =>
  UpdateImage({
    imageIndex,
    incrementOrDecrement: 'increment',
    setActiveDotIndex,
    setImageIndex,
    setImageLoaded,
  });

DecrementImage.propTypes = {
  imageIndex: PropTypes.number.isRequired,
  setActiveDotIndex: PropTypes.func.isRequired,
  setImageIndex: PropTypes.func.isRequired,
  setImageLoaded: PropTypes.func.isRequired,
};
IncrementImage.propTypes = DecrementImage.propTypes;
UpdateImage.propTypes = {
  ...DecrementImage.propTypes,
  incrementOrDecrement: PropTypes.string.isRequired,
};
