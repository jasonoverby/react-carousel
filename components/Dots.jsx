import React from 'react';
import PropTypes from 'prop-types';

import { imageLocations } from '../lib/image-locations';
import { updateImage } from '../lib/update-image';

export const Dots = ({
  activeDotIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  const dots = imageLocations.map((_, i) => (
    <Dot
      activeDotIndex={activeDotIndex}
      index={i}
      key={`dot-${i}`}
      setActiveDotIndex={setActiveDotIndex}
      setImageIndex={setImageIndex}
      setImageLoaded={setImageLoaded}
    />
  ));

  return <div className="dots">{dots}</div>;
};

export const Dot = ({
  activeDotIndex,
  index,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  return (
    <button
      className={index === activeDotIndex ? 'active' : ''}
      // image number in label should not be zero-indexed as that might be confusing
      aria-label={`show image ${index + 1}`}
      onClick={async () => {
        await updateImage({
          updatedImageIndex: index,
          setActiveDotIndex,
          setImageIndex,
          setImageLoaded,
        });
      }}
      type="button"
    >
      <span>â€¢</span>
    </button>
  );
};

Dots.propTypes = {
  activeDotIndex: PropTypes.number.isRequired,
  setActiveDotIndex: PropTypes.func.isRequired,
  setImageIndex: PropTypes.func.isRequired,
  setImageLoaded: PropTypes.func.isRequired,
};
Dot.propTypes = {
  ...Dots.propTypes,
  index: PropTypes.number.isRequired,
};
