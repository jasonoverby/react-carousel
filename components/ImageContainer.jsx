import React from 'react';
import PropTypes from 'prop-types';

import { imageLocations } from '../lib/image-locations';

export const ImageLink = ({ imageIndex, imageLoaded, setCarouselPaused }) => {
  return (
    <a
      aria-label={`Visit
        ${imageLocations[imageIndex].href} (opens in a new tab)`}
      href={imageLocations[imageIndex].href}
      onMouseOver={() => setCarouselPaused(true)}
      // for keyboard navigation
      onFocus={() => setCarouselPaused(true)}
      onMouseOut={() => setCarouselPaused(false)}
      // for keyboard navigation
      onBlur={() => setCarouselPaused(false)}
      rel="noreferrer"
      target="_blank"
    >
      <img
        alt={imageLocations[imageIndex].alt}
        aria-label="This image will change every ten seconds unless the cursor is hovering over it"
        src={imageLocations[imageIndex].src}
        className={imageLoaded ? 'imageLoaded' : ''}
      />
    </a>
  );
};

export const ImageContainer = ({
  imageIndex,
  imageLoaded,
  setCarouselPaused,
}) => (
  <div className="imgContainer">
    <ImageLink
      imageIndex={imageIndex}
      imageLoaded={imageLoaded}
      setCarouselPaused={setCarouselPaused}
    />
  </div>
);

ImageContainer.propTypes = {
  imageIndex: PropTypes.number.isRequired,
  imageLoaded: PropTypes.bool.isRequired,
  setCarouselPaused: PropTypes.func.isRequired,
};
ImageLink.propTypes = ImageContainer.propTypes;
