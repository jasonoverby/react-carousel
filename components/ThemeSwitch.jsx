import React from 'react';
import PropTypes from 'prop-types';

export const ThemeSwitch = ({ checked, toggleTheme }) => {
  const handleKeyPress = (e) => {
    // if user presses spacebar when switch is in focus
    // theme is toggled
    if (e.keyCode === 32) {
      // prevents default browser scroll on spacebar behavior
      e.preventDefault();
      toggleTheme();
    }
  };
  return (
    <>
      <span className="themeIcon">ðŸŒž</span>
      <div className="themeSwitch">
        <label tabIndex="0" onKeyDown={handleKeyPress}>
          <input
            type="checkbox"
            role="switch"
            onChange={toggleTheme}
            checked={checked}
            aria-checked={checked}
          />
          <span className="slider"></span>
        </label>
      </div>
      <span className="themeIcon">ðŸŒœ</span>
    </>
  );
};

ThemeSwitch.propTypes = {
  checked: PropTypes.bool,
  toggleTheme: PropTypes.func,
};
