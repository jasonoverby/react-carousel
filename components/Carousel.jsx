import React, { useEffect, useState } from 'react';

import { ImageContainer } from './ImageContainer';
import { Dots } from './Dots';
import { DecrementImage, IncrementImage } from './ChangeImage';
import { getNewImageIndex, updateImage } from '../lib/utils';

export const Carousel = () => {
  const [imageIndex, setImageIndex] = useState(0);
  const [carouselPaused, setCarouselPaused] = useState(false);
  const [imageLoaded, setImageLoaded] = useState(true);
  const [activeDotIndex, setActiveDotIndex] = useState(0);

  useEffect(() => {
    const intervalId = setInterval(async () => {
      if (!carouselPaused) {
        const updatedImageIndex = getNewImageIndex(imageIndex, 'increment');
        await updateImage({
          setActiveDotIndex,
          setImageIndex,
          setImageLoaded,
          updatedImageIndex,
        });
      }
    }, 10000);
    return () => clearInterval(intervalId);
  });

  return (
    <div className="container">
      <DecrementImage
        imageIndex={imageIndex}
        setActiveDotIndex={setActiveDotIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
      <ImageContainer
        imageIndex={imageIndex}
        imageLoaded={imageLoaded}
        setCarouselPaused={setCarouselPaused}
      />
      <IncrementImage
        imageIndex={imageIndex}
        setActiveDotIndex={setActiveDotIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
      <Dots
        activeDotIndex={activeDotIndex}
        setActiveDotIndex={setActiveDotIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
    </div>
  );
};
