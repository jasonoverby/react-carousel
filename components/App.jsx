import React from 'react';

import { ThemeSwitch } from './ThemeSwitch';
import { Carousel } from './Carousel';

export const App = () => {
  // destructuring useState on import makes it difficult to mock so using React.useState instead
  const [darkTheme, setDarkTheme] = React.useState(false);
  return (
    <main className={darkTheme ? 'darkTheme' : ''}>
      <ThemeSwitch toggleTheme={setDarkTheme} checked={darkTheme} />
      <Carousel />
    </main>
  );
};
