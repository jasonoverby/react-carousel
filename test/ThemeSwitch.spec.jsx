import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { ThemeSwitch } from '../components/ThemeSwitch';

describe('ThemeSwitch', () => {
  const toggleTheme = jest.fn();

  it('renders correctly', () => {
    const checked = false;
    const { asFragment } = render(
      <ThemeSwitch checked={checked} toggleTheme={toggleTheme} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when checked is true', () => {
    const checked = true;
    const { asFragment } = render(
      <ThemeSwitch checked={checked} toggleTheme={toggleTheme} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls toggleTheme when input is clicked', () => {
    const checked = false;
    const { container } = render(
      <ThemeSwitch checked={checked} toggleTheme={toggleTheme} />,
    );

    const switchInput = container.querySelector('#switch');
    fireEvent.click(switchInput);
    expect(toggleTheme).toHaveBeenCalledWith(!checked);
  });
});
