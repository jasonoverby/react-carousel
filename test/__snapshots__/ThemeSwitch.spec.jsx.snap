// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitch renders correctly 1`] = `
<DocumentFragment>
  <span
    class="themeIcon"
  >
    ðŸŒž
  </span>
  <div
    class="themeSwitch"
  >
    <label
      for="switch"
    >
      <input
        aria-checked="false"
        id="switch"
        role="switch"
        type="checkbox"
      />
      <span
        class="slider"
      />
    </label>
  </div>
  <span
    class="themeIcon"
  >
    ðŸŒœ
  </span>
</DocumentFragment>
`;

exports[`ThemeSwitch renders correctly when checked is true 1`] = `
<DocumentFragment>
  <span
    class="themeIcon"
  >
    ðŸŒž
  </span>
  <div
    class="themeSwitch"
  >
    <label
      for="switch"
    >
      <input
        aria-checked="true"
        checked=""
        id="switch"
        role="switch"
        type="checkbox"
      />
      <span
        class="slider"
      />
    </label>
  </div>
  <span
    class="themeIcon"
  >
    ðŸŒœ
  </span>
</DocumentFragment>
`;
